Parts-functions

kristenauriemma opened the pull request:
created parts.js functions and partSearch.handlebars

kristenauriemma committed:
Updated parts.js

created functions createPart, getPartById, searchPartsByName, and getCarsByPartId

kristenauriemma committed:
Update parts.js

kristenauriemma committed:
Created partSearch.handlebars

partSearch.handlebars created under views

kristenauriemma committed:
Updated parts.js

kristenauriemma committed:
Create partRoutes.js

part routes

kristenauriemma committed:
Update partRoutes.js

kristenauriemma committed:
Updated index.js

included parts

kristenauriemma committed:
Update partRoutes.js

kristenauriemma committed:
Update partRoutes.js

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+        manufacturer: partData.manufacturer,
+        sellerId: new ObjectId(partData.sellerId),
+        carIds: partData.carIds.map(id => new ObjectId(id)),
+        dateAdded: new Date()
```
use these as standard:
createdAt: updateAt: new Date().toUTCString()
updateAt: new Date().toUTCString()

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+            throw new Error('Could not add part');
+        }
+        const newId = insertInfo.insertedId;
+        return await getPartById(newId.toString());
```
don't query the database, just return: 
{...newPart, _id: insertInfo.insertedId}

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+        if (!part) throw new NotFoundException('Part not found');
+        return part;
+    } catch (e) {
+        throw new DataBaseException(e.message);
```
use databaseExceptionHandler(e) util function instead, throw databaseexception here will override the notfoundexception 404 return.

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+        const partsCollection = await parts();
+        const parts = await partsCollection.find(
+            { name: { $regex: searchQuery, $options: 'i' } },
+            { projection: { name: 1, price: 1, manufacturer: 1, specifications: 1, dateAdded: 1 } } 
```
I don't think specifications exist, if you gonna add it, make sure its part of the schema (in the insert function). Also use createAt() insteadof dateAdded.  I would just remove the whole projection, as we don't have anything to hide from user.

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+    }
+};
+
+const getCarsByPartId = async (partId) => {
```
I think we need getPartByCarId, as the user usually would search parts entering cars

nuendo2004 commented on src/routes/partRoutes.js:
```diff
@@ -0,0 +1,59 @@
+
+router.get('/:partId', auth, async (req, res, next) => {
+    try {
+        const part = await getPartById(req.params.partId);
```
validate the id

nuendo2004 commented on src/routes/partRoutes.js:
```diff
@@ -0,0 +1,59 @@
+    }
+});
+
+router.get('/search', auth, async (req, res, next) => {
```
use authSafe, we should allow user to search without login

nuendo2004 commented on src/data/parts.js:
```diff
@@ -0,0 +1,82 @@
+    try {
+        const partsCollection = await parts();
+        const parts = await partsCollection.find(
+            { name: { $regex: searchQuery, $options: 'i' } },
```
I would use or to have the search query match manufacture and car make and model as well.
example:
const results = await partsCollection.aggregate([
    // Step 1: Join with the cars collection
    {
        $lookup: {
            from: "cars",                // The collection to join with
            localField: "manufacturer",  // Field from partsCollection
            foreignField: "make",        // Field from cars collection
            as: "carDetails"             // Alias for the joined data
        }
    },
    // Step 2: Unwind the joined data (optional, depending on structure)
    {
        $unwind: {
            path: "$carDetails",
            preserveNullAndEmptyArrays: true // Optional: If you want parts with no car match
        }
    },
    // Step 3: Filter based on the search query across multiple fields
    {
        $match: {
            $or: [
                { name: { $regex: searchQuery, $options: 'i' } },
                { manufacturer: { $regex: searchQuery, $options: 'i' } },
                { "carDetails.make": { $regex: searchQuery, $options: 'i' } },
                { "carDetails.model": { $regex: searchQuery, $options: 'i' } }
            ]
        }
    }
]).toArray();

nuendo2004 commented on src/routes/partRoutes.js:
```diff
@@ -0,0 +1,59 @@
+    try {
+        const { query } = req.query;
+        const parts = await searchPartsByName(query);
+        res.render('partSearch', { results: parts });
```
add user: req.user to the data for all render
res.render('partSearch', { results: parts, user: req.user });

nuendo2004 commented on src/routes/partRoutes.js:
```diff
@@ -0,0 +1,59 @@
+    }
+});
+
+router.use(error);
```
remove this, we already have an error middleware added server level

nuendo2004 commented on src/views/partSearch.handlebars:
```diff
@@ -0,0 +1,41 @@
+    <link rel="stylesheet" href="/public/index.css">
+</head>
+
+<body>
```
remove all html boilerplate(code above), its already added templatewide

kristenauriemma committed:
Update partSearch.handlebars

removed all html boilerplate

kristenauriemma committed:
Updated partRoutes.js

*validated id
*changed auth to authSafe
*added user: req.user to render
*removed error middleware

kristenauriemma committed:
Updated createPart in parts.js

*added {...newPart, _id: insertInfo.insertedId}
*used createdAt and updatedAt

kristenauriemma committed:
Updated parts.js

nuendo2004 committed:
Merge branch 'dev' into parts-functions
